solutions/cf-1713E.mdx

```diff
@@ -0,0 +1,163 @@
+
+We can create a disjoint set union to model this problem. The DSU will have $2n$ nodes where node $i$ represents the state where the operation is performed on the $i$th row and node $i + n$ represents the state where operation is not performed on the $i$th row.
+
+As we are iterating greedily through the array, we can merge ($i$ and $j$) and ($i + n$ and $j + n$) if we don't want to swap. This works because if we swap neither or swap both the arrangement remains the same.
+If we do want to swap we can merge ($i$ and $j + n$) and ($i + n$ and $j$). This works because if we swap exactly one of these, then the arrangement will change. 
+
+After this we can construct the answer by iterating over each cell of the grid. If the cell's row index is equal to its column index (i.e. it's on the diagonal of the grid) it can never change regardless of the flips. If $a[i][j] == a[j][i]$ then we can manually set them (since we didn't merge them in the disjoint set union when comparing them the states might have not been merged). Otherwise, we can check whether $i$ and $j$ are merged in the disjoint set union. If they aren't then we will swap $a[i][j]$ and $a[j][i]$. Otherwise we will keep them the same.
+
+## Implementation
+
+**Time Complexity:** $\mathcal{O}(N^2)$
+
+<LanguageSection>
+
+<CPPSection>
+
+```cpp
+#include <bits/stdc++.h>
+using namespace std;
+
+using vi = vector<int>;
+
+//BeginCodeSnip{DSU}
+struct DSU {
+	vi e;
+	DSU(int n) {
+		e = vi(n, -1);
+	}
+	
+	int get(int x) {
+		if (e[x] < 0) {
+			return x;
+		} else {
+			e[x] = get(e[x]);
+			return e[x];
+		}
+	}
+	
+	bool same(int a, int b) {
+		return get(a) == get(b);
+	}
+	
+	int size(int x) {
+		return -e[get(x)];
+	}
+	
+	bool unite(int x, int y) {
+		x = get(x);
+		y = get(y);
+		if (x == y) {
+			return false;
+		}
+		if (e[x] > e[y]) {
+			swap(x, y);
+		}
+		e[x] += e[y];
+		e[y] = x;
+		return true;
+	}
+};
+//EndCodeSnip
+
+void solve() {
+	int n;
+	cin >> n;
+	int a[n][n];
+	for (int i = 0; i < n; i++) {
+		for (int j = 0; j < n; j++) {
+			cin >> a[i][j];
+		}
+	}
+	
+	DSU dsu(2 * n);
+	
+	for (int i = 0; i < n; i++) {
+		// Only check each pair once and diagnals don't matter
+		for (int j = i + 1; j < n; j++) {
+			// If they are the same we should not merge either way
+			// because it could affect later merges
+			if (a[i][j] == a[j][i]) {
+				continue;
+			} else if (a[i][j] > a[j][i]) {
+				// Swap the elements if the states are not already merged the opposite way
+				if (dsu.get(i) != dsu.get(j)) {
+					dsu.unite(i, j + n);
+					dsu.unite(j, i + n);
+				}
+			} else	{
+				// Keep the elements the same if the states
+				// are not already merged the opposite way
+				if (dsu.get(i) != dsu.get(j + n)) {
+					dsu.unite(i, j);
+					dsu.unite(i + n, j + n);
+				}
+			}
+		}
+	}
+	
+	// Construct the answer
+	int ans[n][n];
```

Comment:
please use a vector

Reply:
vectors are too slow

Reply:
i got inconsistent results when i used a vector
sometimes tle and sometimes ac